base/coreir.jl

```diff
@@ -14,61 +14,75 @@ Core.Const
 """
     struct PartialStruct
         typ
-        undef::BitVector # represents whether a given field may be undefined
+        undefs::Vector{Union{Nothing,Bool}} # represents whether a given field may be undefined
         fields::Vector{Any} # i-th element describes the lattice element for the i-th defined field
     end
 
 This extended lattice element is introduced when we have information about an object's
 fields beyond what can be obtained from the object type. E.g. it represents a tuple where
 some elements are known to be constants or a struct whose `Any`-typed field is initialized
-with `Int` values.
+with value whose type is concrete.
 
 - `typ` indicates the type of the object
-- `undef` records which fields are possibly undefined
-- `fields` holds the lattice elements corresponding to each defined field of the object
-
-If `typ` is a struct, `undef` represents whether the corresponding field of the struct is guaranteed to be
-initialized. For any defined field, there is a corresponding `fields` element which provides information
-about the type of the defined field.
-
-If `typ` is a tuple, the last element of `fields` may be `Vararg`. In this case, it is
-guaranteed that the number of elements in the tuple is at least `length(fields)-1`, but the
-exact number of elements is unknown (`undef` then has a length of `length(fields)-1`).
+- `undefs` records defined-ness of each field
+- `fields` holds the lattice elements corresponding to each field of the object
+
+`fields` corresponds to the fields that `typ` can have.
+If `typ` is a struct that can have `n` fields, then `length(fields) == n`.
+A special case: if `typ` is a variable-length `Tuple`,
+then `length(fields) == datatype_min_ninitialized(typ) + 1`.
+The last element represents the `Vararg` element.
+
+`undefs` is a `Vector{Union{Nothing,Bool}}` with the same length as `fields`, encoding
+the following information about field defined-ness:
+- `undefs[i] === nothing` indicates the corresponding element in `fields` may be undefined
+- `undefs[i] === false` indicates the corresponding element in `fields` is guaranteed to be defined
+- `undefs[i] === true` indicates the corresponding element in `fields` is guaranteed to be undefined
+If `field[i]` is of type `Union{}`, it means the `i`-th field is never be initialized and
+thus never be defined. In this case, `undefs[i]` should always be `true`.
```

Comment:
Just for my understanding, was there any situation that we couldn't represent with the old encoding?

The docs seem to suggest `undefs[i] === true` <-> `fields[i] === Union{}` but maybe there's a corner-case where that's not true?

Reply:
`undefs[i] === true` -> `fields[i] === Union{}` is true, but the reverse isn't. Immutable structs with uninitialized pointer fields can have `undefs[i] === true` but at the same time `fields[i]` can be arbitrary types. We can probably make those `fields[i]` to `Union{}`, but I wanted to avoid encoding field defined-ness information into `fields`.

Reply:
OK, that makes sense. Thanks for the details 👍

BTW, are you required to include anything that might come out of the type, or just what went into it? For example, if I make `PartialStruct(Union{Foo,Bar}, undefs=[...], fields=[...])` where:
```julia
struct Foo
    a::Int
    b::Int
    Foo() = new()
end
struct Bar
    a::Any
    b::Any
    Bar() = new()
end
```

with `rand(Bool) ? Foo() : Bar()` (i.e. both fields undefined) - are the types in `fields` required to include `Int`, since that describes what we might read out of the `isbits` fields?

Or is it still correct to leave `typ` at `Union{}` in that case?

Reply:
At least for now, `PartialStruct` cannot represent objects that are inferred as `Union` types. The expression `(x::PartialStruct).typ` is actually either a `DataType` or a `UnionAll`.

Reply:
The UnionAll equivalent would be `PartialStruct(Foo, ...)` where:
```julia
struct Foo{T}
     a::T
     b::T
     Foo{T}() where T = new{T}()
 end
 ```
 
 `Foo{Int}` is `isbits` and `Foo{Any}` is not